### PR TITLE
[effects] Added delete, export, import preset implementation. Step 4

### PR DESCRIPTION
**Pull request by igorkorsukov**
Resolves: part of https://github.com/audacity/audacity/issues/7761

### CODE REVIEW

File: src/effects/effects_base/internal/effectspresetscontroller.cpp [hunk 1]
```diff
@@ -107,21 +110,129 @@ Ret EffectsPresetsController::saveCurrentAsPreset(const EffectInstanceId& effect
...
+    params = params.AfterFirst(':');
+
+    auto commandId = effect->GetSquashedName(effect->GetSymbol().Internal());
+
+    if (ident != commandId) {
+        // effect identifiers are a sensible length!
+        // must also have some params.
+        std::string msg;
+        if ((params.Length() < 2) || (ident.Length() < 2) || (ident.Length() > 30)) {
+            msg = muse::mtrc("effetcs", "%1: is not a valid presets file.").arg(path.toString()).toStdString();
```
Comment (kryksyh) [inline src/effects/effects_base/internal/effectspresetscontroller.cpp:185]: "effetcs" - a typo

File: src/effects/effects_base/internal/effectspresetscontroller.cpp [hunk 2]
```diff
@@ -107,21 +110,129 @@ Ret EffectsPresetsController::saveCurrentAsPreset(const EffectInstanceId& effect
...
+    }
+
+    EffectSettings* settings = instancesRegister()->settingsById(effectInstanceId);
+    IF_ASSERT_FAILED(settings) {
+        return muse::make_ret(Ret::Code::InternalError);
+    }
+
+    const std::string interactiveTitle = muse::trc("effects", "Import Effect Parameters");
+    io::path_t path = interactive()->selectOpeningFile(QString::fromStdString(interactiveTitle),
+                                                       globalConfiguration()->homePath(),
```
Comment (kryksyh) [inline src/effects/effects_base/internal/effectspresetscontroller.cpp:160]: A minor question: Is there some machinery that can allow us to save the last directory used for each kind of dialog? 
This is a minor thing that always gets me when the app does not remember what directory I've used before.
Imagine that you have downloaded a pack of presets and then you are importing them one by one and every time you have to navigate to the directory. 

The same goes for audio export and any other file operations. Using the last directory for each kind of open dialog would be really helpful.

**Review by kryksyh — APPROVED** [review]

File: src/effects/effects_base/internal/effectspresetscontroller.cpp (hunk 1 above)
Comment (igorkorsukov) [reply src/effects/effects_base/internal/effectspresetscontroller.cpp]: fixed

File: src/effects/effects_base/internal/effectspresetscontroller.cpp (hunk 2 above)
Comment (igorkorsukov) [reply src/effects/effects_base/internal/effectspresetscontroller.cpp]: added last path

File: src/effects/builtin/common/abstracteffectmodel.h [hunk 3]
```diff
@@ -53,17 +53,28 @@ class AbstractEffectModel : public QObject, public muse::async::Asyncable
...
     }
 
     template<typename T>
     T& mutSettings()
     {
-        auto pSettings = settings()->cast<T>();
-        assert(pSettings);
-        return *pSettings;
+        EffectSettings* s = this->settings();
+        assert(s);
```
Comment (kryksyh) [inline src/effects/builtin/common/abstracteffectmodel.h:70]: This assert is no longer needed

File: src/effects/effects_base/effecterrors.h [hunk 4]
```diff
@@ -31,6 +35,10 @@ inline muse::Ret make_ret(Err e)
         return muse::Ret(retCode, muse::trc("effects", "No audio selected"));
     case Err::EffectProcessFailed:
         return muse::Ret(retCode, muse::trc("effects", "Applying effect failed"));
+    case Err::PresetNotValid:
+        return muse::Ret(retCode, muse::trc("effects", "Not valid presets file"));
+    case Err::PresetMissMatch:
+        return muse::Ret(retCode, muse::trc("effects", "Preset miss match"));
```
Comment (kryksyh) [inline src/effects/effects_base/effecterrors.h:41]: Typo: it should be 'mismatch'

File: src/effects/builtin/common/abstracteffectmodel.h (hunk 3 above)
Comment (igorkorsukov) [reply src/effects/builtin/common/abstracteffectmodel.h:70]: I decided to keep it for `mutSettings`

File: src/effects/effects_base/effecterrors.h (hunk 4 above)
Comment (igorkorsukov) [reply src/effects/effects_base/effecterrors.h]: fixed

**Pull request merged**